### PR TITLE
fix: use a factory to create the itemTags keys

### PR DESCRIPTION
**Pull request by spaenleh**
In this PR I propose a way to organize query keys with a factory.
The many item tags query now contains all the item ids instead of a hash, which allows us to invalidate all many item tags caches that handle a certain item id when that item tags are updated using the single call.
React-query does not look to have a limit on the size of the query keys so this should be fine.

### CODE REVIEW

File: src/config/keys.ts [hunk 1]
```diff
@@ -54,7 +54,20 @@ export const buildManyItemMembershipsKey = (ids?: UUID[]) => [
 export const buildItemLoginKey = (id?: UUID) => [ITEMS_KEY, id, 'login'];
 export const TAGS_KEY = 'tags';
 export const ITEM_TAGS_KEY = 'itemTags';
+export const itemTagsKeys = {
```
Comment (pyphilia) [inline src/config/keys.ts:57]: LGTM but should we have a better name for the factory?

File: src/mutations/itemTag.ts [hunk 2]
```diff
@@ -65,12 +67,14 @@ export default (queryClient: QueryClient, queryConfig: QueryClientConfig) => {
...
-      const itemKey = buildItemTagsKey(id);
+      const itemKey = itemTagsKeys.singleId(id);
       queryClient.setQueryData(itemKey, context.itemTags);
       notifier?.({ type: deleteItemTagRoutine.FAILURE, payload: { error } });
     },
     onSettled: (_data, _error, { id }) => {
-      queryClient.invalidateQueries(buildItemTagsKey(id));
+      queryClient.invalidateQueries(itemTagsKeys.singleId(id));
+      // invalidate any "many" query that contains the id we modified
+      queryClient.invalidateQueries(itemTagsKeys.manyIds([id]));
```
Comment (pyphilia) [inline src/mutations/itemTag.ts:77]: Would it make sense to have a "util" of some sort to handle an invalidation of single and many given an id?  Not sure how

**Review by pyphilia — APPROVED** [review]
LGTM, but I'll let you rebase for the refactor :P

File: src/config/keys.ts [hunk 3]
```diff
@@ -54,7 +54,20 @@ export const buildManyItemMembershipsKey = (ids?: UUID[]) => [
...
+  all: [ITEMS_KEY, ITEM_TAGS_KEY] as const,
+  many: () => [...itemTagsKeys.all, 'many'] as const,
+  single: () => [...itemTagsKeys.all, 'single'] as const,
```
Comment (dialexo) [inline src/config/keys.ts:60]: Not sure if these 3 helper prefix factories should be part of the public API, consumers probably don't want to use these. Consider using a class with private methods, or non-exported functions.

```suggestion
```

**Review by dialexo — CHANGES_REQUESTED** [review]
WDYT? Does the following make sense?

File: src/config/keys.ts (hunk 1 above)
Comment (spaenleh) [reply src/config/keys.ts:57]: What name were you thinking of ?

File: src/mutations/itemTag.ts (hunk 2 above)
Comment (spaenleh) [reply src/mutations/itemTag.ts:77]: I guess we could make a util that would invalidate single and many queries relating to a given id, not sure of the benefit though...

File: src/config/keys.ts (hunk 3 above)
Comment (dialexo) [reply src/config/keys.ts:60]: I see, I didn't know React Query supported key prefix access

**Review by dialexo — APPROVED** [review]

**Pull request merged**